### DISCUSSION TITLE
Custom datasource via REST API

### DISCUSSION THREAD

**Discussion by addisonElliott** [directus/directus#9093] [Feature Requests]
It would be nice if Directus had the capability to interact with a custom datasource via REST API. My company has a few databases we'd like to be able to interact with. Our databases are MongoDB and they’re hosted in different places.

My vision is that revision control would still be handled in Directus. User/role management still done via Directus too. There'd be a place where you specify the schema of the REST API.

This does deviate away from what Directus is meant to do, but my use case is that I have existing datasources that I want to connect with Directus. My goal is to have a central admin dashboard where you can interact with Directus **and** other datasources.

**Comment by joselcvarela** [comment]
Hello @addisonElliott 
Here I also have talked about this https://github.com/directus/directus/discussions/9009#discussioncomment-1518527

But your solution would be even further and that's cool.
The data seems to be somehow easy to handle, but the relationship with images and filesystem or external source could be less easier.

**Comment by addisonElliott** [comment]
Yeah, it's a similar concept as having the capability to migrate a datasource into Directus. The big difference is that I'm envisioning having "virtual collections" where the data isn't actually stored in Directus. Directus interacts with this data similarly to real collections except it will utilize a REST API to do CRUD actions.

This idea came from a React project called [react-admin](https://github.com/marmelab/react-admin). The purpose of the project is to provide an admin dashboard frontend in order to alter data from a datasource.

On a high-level, my thoughts are that any number of "virtual collections" can be defined that act just like real collections in Directus. A "data provider" class is provided for each virtual collection. This would just be a class instance that has logic for doing CRUD operations.

I think you're right that there are some edge cases that will prove difficult. Handling relationships between data in these virtual collections would be interesting as an example.

I'm not quite convinced this passes the 80/20 requirement, but having the capability to do this via an extension would be sufficient for me.

**Reply by benhaynes** [reply]
This seems very similar to the idea of adding support for SQL Views into Directus, right?

https://github.com/directus/directus/discussions/3097

**Reply by rijkvanzanten** [reply]
@benhaynes No. A view is a different way of interfacing with the same dataset in the same datastore. This is asking for connecting multiple datastores to the same Directus instance 👍🏻

**Comment by addisonElliott** [comment]
@benhaynes @rijkvanzanten I read somewhere (not sure where) about plans for Directus 10 moving towards abstracting the data provider away to _potentially_ support NoSQL. Would/could this idea play into that?

**Reply by rijkvanzanten** [reply]
Supporting NoSQL is a pretty different discussion from supporting multiple datastores simultaneously. It's a little premature for me to make any promises about what a next major version of the platform might look like (seeing v9 isn't even out the door yet) 🙂

**Reply by addisonElliott** [reply]
Yep, totally understand. Not hoping for any commitment from you. Especially since I imagine multiple datastores is deviating from the main purpose of Directus.

I bring this up because my (tentative) plan is to fork Directus and add this functionality for use at my company. I am scheduled to work on an internal admin dashboard early next year and I hope to do it in such a way that it can eventually be an extension or a core part of Directus. Once (or if) I have anything appreciable to present, I'll post it here for thoughts.

**Reply by rijkvanzanten** [reply]
Curious to hear what you'll come up with. That's no small feat

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
